tests/unit_tests/scheduler/test_lsf_driver.py

```diff
@@ -32,61 +36,522 @@
+    mock_bsub, caplog, tmpdir, monkeypatch
+):
+    monkeypatch.chdir(tmpdir)
+    lsf_driver = LSFDriver(None)
+    lsf_driver = LSFDriver(queue_options=[("BSUB_CMD", tmpdir / "mock_bsub")])
+    lsf_driver._max_attempt = 3
+    lsf_driver._retry_sleep_period = 0
+
+    mock_realization_state = MockRealizationState()
+    mock_realization_state.realization.job_script = "exit.sh"
+
+    with pytest.raises(RuntimeError, match="Maximum number of submit errors exceeded"):
+        await lsf_driver.submit(mock_realization_state)
+
+    job_paths = Path("job_paths").read_text(encoding="utf-8").strip().split("\n")
+
+    # should try command 3 times before exiting
+    assert len(job_paths) == lsf_driver._max_attempt
+
+    output = caplog.text
+    assert len(re.findall("bsub returned non-zero exitcode: 1", output)) == 3
+
+
+@pytest.mark.asyncio
+async def test_that_submit_fails_when_bsub_returns_badly_formated_response(
+    mock_bsub, caplog, tmpdir, monkeypatch
+):
+    monkeypatch.chdir(tmpdir)
+    lsf_driver = LSFDriver(queue_options=[("BSUB_CMD", tmpdir / "mock_bsub")])
+    lsf_driver._max_attempt = 3
+    lsf_driver._retry_sleep_period = 0
+
+    mock_realization_state = MockRealizationState()
+    mock_realization_state.realization.job_script = "garbled_return.sh"
+
+    with pytest.raises(RuntimeError, match="Maximum number of submit errors exceeded"):
+        await lsf_driver.submit(mock_realization_state)
+
+    job_paths = Path("job_paths").read_text(encoding="utf-8").strip().split("\n")
+
+    # should try command 3 times before exiting
+    assert len(job_paths) == 3
+
+    output = caplog.text
+    print(f"{output=}")
+    assert len(re.findall("Could not parse lsf id from", output)) == 3
+
+
+@pytest.mark.asyncio
+async def test_submit_success(mock_bsub, caplog, tmpdir, monkeypatch):
+    caplog.set_level(logging.DEBUG)
+    monkeypatch.chdir(tmpdir)
+    lsf_driver = LSFDriver(queue_options=[("BSUB_CMD", tmpdir / "mock_bsub")])
+    lsf_driver._max_attempt = 3
+
+    mock_realization_state = MockRealizationState()
+    mock_realization_state.realization.job_script = "valid_script.sh"
+
+    await lsf_driver.submit(mock_realization_state)
+
+    job_paths = Path("job_paths").read_text(encoding="utf-8").strip().split("\n")
+    assert len(job_paths) == 1
+    output = caplog.text
+    assert re.search("Submitted job.*and got LSF JOBID", output)
+    assert re.search("submitted to default queue", output)
+    assert mock_realization_state.current_state.id == "PENDING"
+
+
+@pytest.mark.asyncio
+async def test_that_poll_statuses_does_not_run_again_while_already_polling(
+    mock_bjobs, caplog, tmpdir, monkeypatch
+):
+    monkeypatch.chdir(tmpdir)
+
+    lsf_driver = LSFDriver(None)
+    lsf_driver._currently_polling = True
+
+    await lsf_driver.poll_statuses()
+
+    # Should not call bjobs
+    assert not Path("bjobs_logs").exists()
+
+    output = caplog.text
+    assert output == ""
```

Comment:
This line is perhaps a little fragile, as it does not assert our wanted behaviour, only what we log, which we might want to change for other reasons later?